### PR TITLE
Feature/gait-schedular

### PR DESCRIPTION
**Pull request by TimBuckers**
- Schedule gait directly if queue is empty.
- Schedule gait directly after an other gait.
- Above options can also be schedule with offset
example gaits:
https://github.com/project-march/march-iv/tree/test/fake-gaits-for-scheduling

### CODE REVIEW

File: march_gait_scheduler/src/Scheduler.cpp [hunk 1]
```diff
@@ -50,10 +50,23 @@ ros::Time Scheduler::getStartTime(ros::Duration offset)
     }
   }
 }
+bool Scheduler::lastScheduledGaitInProgress() {
+  ros::Time currentTime = ros::Time::now();
+  if(currentTime < this->startTimeLastGait){
+    return false;
+  }
+  return true;
+}
```
Comment (Ishadijcks) [inline march_gait_scheduler/src/Scheduler.cpp]:
```
  return ros::Time::now() < this->startTimeLastGait
```

**Comment by Ishadijcks** [comment]
Not needed to use fake gaits as of https://github.com/project-march/gait-files
and https://github.com/project-march/march-iv/pull/215

**Comment by Ishadijcks** [comment]
- Sending an empty goal crashes the gait scheduler. I believe in the `lastScheduledGaitInProgress()`

**Comment by Ishadijcks** [comment]
Without simulation:

```
rostopic pub /march/gait/perform/goal march_shared_resources/GaitNameActionGoal "header:
  seq: 0
  stamp:
    secs: 0
    nsecs: 0
  frame_id: ''
goal_id:
  stamp:
    secs: 0
    nsecs: 0
  id: ''
goal:
  name: 'walk'
  subgait_name: 'right_swing'" 
```
```
Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)
```

File: march_gait_scheduler/include/march_gait_scheduler/Scheduler.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+#include <control_msgs/FollowJointTrajectoryActionGoal.h>
+#include <march_shared_resources/GaitGoal.h>
+#include <ros/duration.h>
+#include <ros/ros.h>
+#include <trajectory_msgs/JointTrajectory.h>
+
+class Scheduler
+{
+  const march_shared_resources::GaitGoal* lastGaitGoal = nullptr;
+  ros::Time startTimeLastGait;
```
Comment (Ishadijcks) [inline march_gait_scheduler/include/march_gait_scheduler/Scheduler.h]: Time is redundant with the Type

File: march_gait_scheduler/include/march_gait_scheduler/Scheduler.h [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  bool lastScheduledGaitInProgress();
+  static trajectory_msgs::JointTrajectory setStartTimeGait(trajectory_msgs::JointTrajectory trajectory, ros::Time time);
+
+public:
+  /**
+   * A gait is succeeded this duration before its actual completion.
+   */
+  ros::Duration GAIT_SUCCEEDED_OFFSET = ros::Duration(0.2);
+  ros::Time getEndTimeCurrentGait();
+  control_msgs::FollowJointTrajectoryGoal scheduleTrajectory(const march_shared_resources::GaitGoal* trajectory,
```
Comment (Ishadijcks) [inline march_gait_scheduler/include/march_gait_scheduler/Scheduler.h]: Is trajectory the correct name for the parameter?

File: march_gait_scheduler/src/GaitSchedulerNode.cpp [hunk 4]
```diff
@@ -1,124 +1,113 @@
...
-actionlib_msgs::GoalStatus trajectory_status;
-
-/**
- * To determine if a goal execution is finished and isn't in progress anymore.
- * @param status the status
- * @return true if the execution is finished if still in progress return false.
- */
-bool statusIsTerminal(const actionlib_msgs::GoalStatus& status)
+typedef actionlib::SimpleActionServer<march_shared_resources::GaitAction> ScheduleGaitActionServer;
+ScheduleGaitActionServer* schedule_gait_action_server;
```
Comment (Ishadijcks) [inline march_gait_scheduler/src/GaitSchedulerNode.cpp]: Please stick to CamelCase

File: march_gait_scheduler/src/GaitSchedulerNode.cpp [hunk 5]
```diff
@@ -1,124 +1,113 @@
...
-void TrajectoryExecutionStatusCallback(const control_msgs::FollowJointTrajectoryActionFeedbackConstPtr& status_message)
-{
-  trajectory_status = actionlib_msgs::GoalStatus();
-  trajectory_status.status = status_message->status.status;
+  while (schedule_gait_action_server->isActive())
+  {
+    ros::Rate r(100);
+    r.sleep();
+    ROS_DEBUG_THROTTLE(1, "Waiting on gait execution");
+  }
```
Comment (Ishadijcks) [inline march_gait_scheduler/src/GaitSchedulerNode.cpp]: Why do you sleep here?

File: march_gait_scheduler/src/GaitSchedulerNode.cpp [hunk 6]
```diff
@@ -1,124 +1,113 @@
...
+  return;
 }
 
 int main(int argc, char** argv)
 {
   ros::init(argc, argv, "gait_scheduler_node");
   ros::NodeHandle n;
-
   ros::Rate rate(100);
```
Comment (Ishadijcks) [inline march_gait_scheduler/src/GaitSchedulerNode.cpp]: 2 rates defined?

File: march_gait_scheduler/src/GaitSchedulerNode.cpp [hunk 7]
```diff
@@ -1,124 +1,113 @@
...
-  //  retrieving planning group name (initialized in launch file) from parameter server
-  n.getParam("/control_node/planning_group", planning_group);
 
-  //  determining which planning group is used. If test_joint, topic name for ServerFollowJoint is changed
+  scheduler = new Scheduler();
 
-  joint_trajectory_pub =
-      n.advertise<control_msgs::FollowJointTrajectoryActionGoal>(TopicNames::follow_joint_trajectory_execution, 1000);
+  followJointTrajectoryAction = new actionlib::SimpleActionClient<control_msgs::FollowJointTrajectoryAction>(
+      "/march/trajectory_controller/follow_joint_trajectory", true);
```
Comment (Ishadijcks) [inline march_gait_scheduler/src/GaitSchedulerNode.cpp]: Set auto_start to false and call it later https://answers.ros.org/question/107126/actionlib-auto_start-parameter/

File: march_gait_scheduler/src/GaitSchedulerNode.cpp [hunk 8]
```diff
@@ -1,124 +1,113 @@
...
-  ros::Subscriber joint_trajectory_status_sub =
-      n.subscribe(TopicNames::follow_joint_trajectory_execution_states, 1000, TrajectoryExecutionStatusCallback);
-
-  ServerFollowJoint server_follow_joint_trajectory(
-      n, ActionNames::schedule_gait,
-      boost::bind(&executeFollowJointTrajectory, _1, &server_follow_joint_trajectory), false);
-  server_follow_joint_trajectory.start();
-
-  while (ros::ok())
+  ROS_INFO("Wait on joint trajectory action server");
```
Comment (Ishadijcks) [inline march_gait_scheduler/src/GaitSchedulerNode.cpp]:
```suggestion
  ROS_DEBUG("Wait on joint trajectory action server");
```

File: march_gait_scheduler/src/GaitSchedulerNode.cpp [hunk 9]
```diff
@@ -1,124 +1,113 @@
...
-    rate.sleep();
-    ros::spinOnce();
+    ROS_INFO("Connected to joint trajectory action server");
   }
+  else
+  {
+    ROS_ERROR("Not connected to joint trajectory action server");
+  }
+
+  schedule_gait_action_server =
```
Comment (Ishadijcks) [inline march_gait_scheduler/src/GaitSchedulerNode.cpp]: CamelCase

File: march_gait_scheduler/src/Scheduler.cpp [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+ros::Time Scheduler::getEndTimeCurrentGait()
+{
+  if (this->lastGaitGoal != nullptr && this->lastGaitGoal != NULL)
+  {
+    ros::Time endTime = this->startTimeLastGait;
+    ROS_DEBUG_THROTTLE(1, "startTimeLastGait: %f", this->startTimeLastGait.toSec());
+    endTime += this->lastGaitGoal->current_subgait.duration;
+    ROS_DEBUG_THROTTLE(1, "this->lastGaitGoal->current_subgait.duration: %f",
```
Comment (Ishadijcks) [inline march_gait_scheduler/src/Scheduler.cpp]: Do these debugs still need to exist?

File: march_gait_scheduler/src/Scheduler.cpp [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+                       this->lastGaitGoal->current_subgait.duration.toSec());
+    return endTime;
+  }
+  return ros::Time::now();
+}
+
+ros::Time Scheduler::getStartTime(ros::Duration offset)
+{
+  ros::Time possibleStartingTime = getEndTimeCurrentGait();
```
Comment (Ishadijcks) [inline march_gait_scheduler/src/Scheduler.cpp]: Rename variable to something like endTimeCurrentGait

File: march_gait_scheduler/src/Scheduler.cpp [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    else
+    {
+      ROS_WARN("Negative offset is ignored, because no gait is currently being "
+               "executed");
+      return currentTime;
+    }
+  }
+  else
+  {
+    if (possibleStartingTime + offset < currentTime)
```
Comment (Ishadijcks) [inline march_gait_scheduler/src/Scheduler.cpp]: This doesn't guarantee that the offset is negative I think?

File: march_gait_scheduler/src/Scheduler.cpp [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+  }
+  return ros::Time::now();
+}
+
+ros::Time Scheduler::getStartTime(ros::Duration offset)
+{
+  ros::Time possibleStartingTime = getEndTimeCurrentGait();
+  ros::Time currentTime = ros::Time::now();
+
+  if (currentTime > possibleStartingTime)
```
Comment (Ishadijcks) [inline march_gait_scheduler/src/Scheduler.cpp]: Reuse this function here? `lastScheduledGaitInProgress()`

File: march_gait_scheduler/src/Scheduler.cpp [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+}
+bool Scheduler::lastScheduledGaitInProgress()
+{
+  return ros::Time::now() >= this->startTimeLastGait;
+}
+
+control_msgs::FollowJointTrajectoryGoal Scheduler::scheduleTrajectory(const march_shared_resources::GaitGoal* goal,
+                                                                      ros::Duration offset)
+{
+  ROS_DEBUG("start time lastgait: %f", this->startTimeLastGait.toSec());
```
Comment (Ishadijcks) [inline march_gait_scheduler/src/Scheduler.cpp]: Good debug statement but make it consistent with the others
```suggestion
  ROS_DEBUG("Start time last gait: %f", this->startTimeLastGait.toSec());
```

File: march_gait_scheduler/src/Scheduler.cpp [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+  ROS_DEBUG("Current time: %f", ros::Time::now().toSec());
+
+  if (!lastScheduledGaitInProgress())
+  {
+    throw std::runtime_error("There is already a gait scheduled in the future.");
+  }
+  ros::Time startingTime = getStartTime(offset);
```
Comment (Ishadijcks) [inline march_gait_scheduler/src/Scheduler.cpp]: startTime?

File: march_gait_scheduler/src/Scheduler.cpp [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+  trajectory_msgs::JointTrajectory trajectory = setStartTimeGait(goal->current_subgait.trajectory, startingTime);
+  control_msgs::FollowJointTrajectoryGoal trajectoryMsg;
+  trajectoryMsg.trajectory = trajectory;
+
+  // Update all variables
```
Comment (Ishadijcks) [inline march_gait_scheduler/src/Scheduler.cpp]: All?

File: march_gait_scheduler/src/GaitSchedulerNode.cpp (hunk 7 above)
Comment (TimBuckers) [reply march_gait_scheduler/src/GaitSchedulerNode.cpp]: This is the case for an action server, this is a client. The client has to wait until the server is online.

File: march_gait_scheduler/src/GaitSchedulerNode.cpp (hunk 5 above)
Comment (TimBuckers) [reply march_gait_scheduler/src/GaitSchedulerNode.cpp]: this way you are not wasting CPU power, by executing this loop all the time

File: march_gait_scheduler/src/Scheduler.cpp (hunk 10 above)
Comment (TimBuckers) [reply march_gait_scheduler/src/Scheduler.cpp]: removed

File: march_gait_scheduler/src/Scheduler.cpp (hunk 13 above)
Comment (TimBuckers) [reply march_gait_scheduler/src/Scheduler.cpp]: lastScheduledGaitInProgress was a misleading name, I fixed that

File: march_gait_scheduler/src/Scheduler.cpp (hunk 12 above)
Comment (TimBuckers) [reply march_gait_scheduler/src/Scheduler.cpp]: It does with the first if combined

**Pull request merged**